### ISSUE TITLE
Java exception IndexOutOfBoundsException

### ISSUE THREAD

**Issue by amenity-green-broker** [zeljkoobrenovic/sokrates#53]
Hello

I am trying to use Sokrates to analyze a company codebase, but I keep getting the same Java exception in one way or another:
```
INFO:  - saving source code cache for the unit_duplicatesfragments
java.lang.IndexOutOfBoundsException: toIndex = 46
        at java.base/java.util.AbstractList.subListRangeCheck(AbstractList.java:509)
        at java.base/java.util.ArrayList.subList(ArrayList.java:1191)
        at nl.obren.sokrates.reports.dataexporters.DataExporter.lambda$saveDuplicateFragmentFiles$38(DataExporter.java:829)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1597)
        at nl.obren.sokrates.reports.dataexporters.DataExporter.lambda$saveDuplicateFragmentFiles$40(DataExporter.java:823)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1597)
        at nl.obren.sokrates.reports.dataexporters.DataExporter.saveDuplicateFragmentFiles(DataExporter.java:813)
        at nl.obren.sokrates.reports.dataexporters.DataExporter.exportSourceFile(DataExporter.java:559)
        at nl.obren.sokrates.reports.dataexporters.DataExporter.saveData(DataExporter.java:117)
        at nl.obren.sokrates.cli.CommandLineInterface.generateReports(CommandLineInterface.java:547)
        at nl.obren.sokrates.cli.CommandLineInterface.generateReports(CommandLineInterface.java:298)
        at nl.obren.sokrates.cli.CommandLineInterface.run(CommandLineInterface.java:128)
        at nl.obren.sokrates.cli.CommandLineInterface.main(CommandLineInterface.java:75)
```
I am not able to share the full error log, but I can give you the anonymized bit and pieces you want. 

The error seems to be related to the duplication analysis. The same error occurs if I leave out dependency analysis, but not if I leave out both dependency and duplication. It will take several hours to check without duplication and dependency, but I am running that test now. The duplication analysis seems to run, but reporting fails:

`INFO: Analysing duplication...
nov. 24, 2024 7:13:20 PM nl.obren.sokrates.common.utils.ProcessingStopwatch start
INFO: Starting analysis/duplication/extracting blocks
nov. 24, 2024 7:13:27 PM nl.obren.sokrates.common.utils.ProcessingStopwatch end
INFO: Done 'analysis/duplication/extracting blocks' in 7278ms
nov. 24, 2024 7:13:27 PM nl.obren.sokrates.common.utils.ProcessingStopwatch start
INFO: Starting analysis/duplication/getting number of duplicated lines
nov. 24, 2024 7:13:27 PM nl.obren.sokrates.common.utils.ProcessingStopwatch end
INFO: Done 'analysis/duplication/getting number of duplicated lines' in 158ms
nov. 24, 2024 7:13:27 PM nl.obren.sokrates.common.utils.ProcessingStopwatch start
INFO: Starting analysis/duplication/finding duplicated units
nov. 24, 2024 7:13:28 PM nl.obren.sokrates.common.utils.ProcessingStopwatch end
INFO: Done 'analysis/duplication/finding duplicated units' in 178ms
nov. 24, 2024 7:13:28 PM nl.obren.sokrates.common.utils.ProcessingStopwatch start
INFO: Starting analysis/duplication/consolidating duplicates
nov. 24, 2024 7:13:29 PM nl.obren.sokrates.common.utils.ProcessingStopwatch end
INFO: Done 'analysis/duplication/consolidating duplicates' in 1411ms
nov. 24, 2024 7:13:29 PM nl.obren.sokrates.common.utils.ProcessingStopwatch start
INFO: Starting analysis/duplication/counting duplicates
nov. 24, 2024 7:13:36 PM nl.obren.sokrates.common.utils.ProcessingStopwatch end
INFO: Done 'analysis/duplication/counting duplicates' in 6847ms
nov. 24, 2024 7:13:36 PM nl.obren.sokrates.sourcecode.analysis.AnalysisUtils detailedInfo
INFO: 26290.73s           - found 699519 duplicates (221134 duplicated lines vs. 1061175 cleaned lines) in 3914 files
nov. 24, 2024 7:13:36 PM nl.obren.sokrates.cli.CommandLineInterface setDetailedText
INFO:   - found 699519 duplicates (221134 duplicated lines vs. 1061175 cleaned lines) in 3914 files
nov. 24, 2024 7:13:42 PM nl.obren.sokrates.sourcecode.analysis.AnalysisUtils detailedInfo
INFO: 26296.98s           - per component:primary
nov. 24, 2024 7:13:42 PM nl.obren.sokrates.cli.CommandLineInterface setDetailedText
INFO:   - per component:primary
`

Again, I can share more if needed after anonymized and names removed.

**Comment by amenity-green-broker** [comment]
Hello

Thank you for the quick reply and update. Running the newly compiled client results in the follow exception:
`ov 25, 2024 10:49:52 AM nl.obren.sokrates.cli.CommandLineInterface setDetailedText
INFO: Unit size per component:
java.lang.NullPointerException
	at nl.obren.sokrates.sourcecode.SourceFile.equals(SourceFile.java:223)
	at nl.obren.sokrates.sourcecode.aspects.LogicalDecomposition.lambda$isInScope$3(LogicalDecomposition.java:154)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at nl.obren.sokrates.sourcecode.aspects.LogicalDecomposition.lambda$isInScope$4(LogicalDecomposition.java:153)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at nl.obren.sokrates.sourcecode.aspects.LogicalDecomposition.isInScope(LogicalDecomposition.java:153)
	at nl.obren.sokrates.sourcecode.units.UnitUtils.lambda$getUnitSizeDistributionPerComponent$4(UnitUtils.java:61)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:178)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1708)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
	at nl.obren.sokrates.sourcecode.units.UnitUtils.lambda$getUnitSizeDistributionPerComponent$6(UnitUtils.java:61)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at nl.obren.sokrates.sourcecode.units.UnitUtils.getUnitSizeDistributionPerComponent(UnitUtils.java:60)
	at nl.obren.sokrates.sourcecode.analysis.files.UnitsAnalyzer.analyze(UnitsAnalyzer.java:79)
	at nl.obren.sokrates.sourcecode.analysis.CodeAnalyzer.analyze(CodeAnalyzer.java:79)
	at nl.obren.sokrates.cli.CommandLineInterface.generateReports(CommandLineInterface.java:544)
	at nl.obren.sokrates.cli.CommandLineInterface.generateReports(CommandLineInterface.java:298)
	at nl.obren.sokrates.cli.CommandLineInterface.run(CommandLineInterface.java:128)
	at nl.obren.sokrates.cli.CommandLineInterface.main(CommandLineInterface.java:75)
`

I can not tell if this is related to the changes or another error. I ran the compiler on Linux Mint with newly installed Maven and used the `cli-1.0-jar-with-dependencies.jar.` I ran the command `java -Xmx20g -jar cli-1.0-jar-with-dependencies.jar generateReports`

**Comment by amenity-green-broker** [comment]
As a test, I ran with openjdk-17-jre instead of openjdk-21-jre and I did not get either exceptions, and the analysis complete (duplication, no dependency). The dependency analysis takes multiple hours on our repository, so I will start it but can not report back soon. I do not know if the original error also came from different jre version, or if you already know about differences between the version. I am happy to provide additional information on what I ran if it is helpful
